drivers/dac/dac_max22017.c

```diff
@@ -0,0 +1,636 @@
+	*value = sys_be16_to_cpu(*value);
+
+	return 0;
+}
+
+static int max22017_reg_write(const struct device *dev, uint8_t addr, uint16_t value)
+{
+	uint8_t crc;
+	size_t crc_len = 0;
+	const struct max22017_config *config = dev->config;
+	const struct max22017_data *data = dev->data;
+
+	addr = FIELD_PREP(MAX22017_SPI_TRANS_ADDR, addr) | FIELD_PREP(MAX22017_SPI_TRANS_DIR, 0);
+	value = sys_cpu_to_be16(value);
+
+	if (data->crc_enabled) {
+		uint8_t crc_in[] = {addr, ((uint8_t *)&value)[0], ((uint8_t *)&value)[1]};
+
+		crc_len = 1;
+		crc = crc8(crc_in, 3, MAX22017_CRC_POLY, 0, true);
+	}
+
+	const struct spi_buf buf[] = {
+		{
+			.buf = &addr,
+			.len = 1,
+		},
+		{
+			.buf = &value,
+			.len = 2,
+		},
+		{
+			.buf = &crc,
+			.len = crc_len,
+		},
+	};
+
+	struct spi_buf_set tx = {
+		.buffers = buf,
+		.count = ARRAY_SIZE(buf),
+	};
+
+	return spi_write_dt(&config->spi, &tx);
+}
+
+static int max22017_channel_setup(const struct device *dev,
+				  const struct dac_channel_cfg *channel_cfg)
+{
+	int ret;
+	uint32_t ao_cnfg, gen_cnfg;
+	uint8_t chan = channel_cfg->channel_id;
+	const struct max22017_config *config = dev->config;
+	struct max22017_data *data = dev->data;
+
+	if (chan > config->nchannels - 1) {
+		LOG_ERR("Unsupported channel %d", chan);
+		return -ENOTSUP;
+	}
+
+	if (channel_cfg->resolution != config->resolution) {
+		LOG_ERR("Unsupported resolution %d", chan);
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->lock, K_FOREVER);
+	ret = max22017_reg_read(dev, MAX22017_AO_CNFG_OFF, &ao_cnfg);
+	if (ret) {
+		goto fail;
+	}
+
+	ao_cnfg |= FIELD_PREP(MAX22017_AO_CNFG_AO_EN, BIT(chan));
+
+	if (!config->latch_mode[chan]) {
+		ao_cnfg |= FIELD_PREP(MAX22017_AO_CNFG_AO_LD_CNFG, BIT(chan));
+	}
+
+	if (config->polarity_mode[chan]) {
+		ao_cnfg |= FIELD_PREP(MAX22017_AO_CNFG_AO_UNI, BIT(chan));
+	}
+
+	if (config->dac_mode[chan]) {
+		ao_cnfg |= FIELD_PREP(MAX22017_AO_CNFG_AO_MODE, BIT(chan));
+	}
+
+	ret = max22017_reg_write(dev, MAX22017_AO_CNFG_OFF, ao_cnfg);
+	if (ret) {
+		goto fail;
+	}
+
+	ret = max22017_reg_read(dev, MAX22017_GEN_CNFG_OFF, &gen_cnfg);
+	if (ret) {
+		goto fail;
+	}
+
+	if (config->ovc_mode[chan]) {
+		gen_cnfg |= FIELD_PREP(MAX22017_GEN_CNFG_OVC_CNFG, BIT(chan));
+		/* Over current shutdown mode */
+		if (config->ovc_mode[chan] == 2) {
+			gen_cnfg |= FIELD_PREP(MAX22017_GEN_CNFG_OVC_SHDN_CNFG, BIT(chan));
+		}
+	}
+
+	ret = max22017_reg_write(dev, MAX22017_GEN_CNFG_OFF, gen_cnfg);
+fail:
+	k_sem_give(&data->lock);
+	return ret;
+}
+
+static int max22017_write_value(const struct device *dev, uint8_t channel, uint32_t value)
+{
+	int ret;
+	uint32_t ao_sta;
+	const struct max22017_config *config = dev->config;
+	struct max22017_data *data = dev->data;
+
+	if (channel > config->nchannels - 1) {
+		LOG_ERR("unsupported channel %d", channel);
+		ret = -ENOTSUP;
+		goto fail;
+	}
+
+	if (value >= (1 << config->resolution)) {
+		LOG_ERR("Value %d out of range", value);
+		ret = -EINVAL;
+		goto fail;
+	}
```

Comment:
These two failure conditions occur before taking the semaphore and shouldn't give it back.